src/bin/str.rs

```diff
@@ -89,88 +132,92 @@ define_language! {
+            | Lang::At(_)
+            | Lang::SubStr(_)
+            | Lang::FromInt(_) => Type::String,
+            Lang::NVar(_)
+            | Lang::Len(_)
+            | Lang::ToInt(_)
+            | Lang::IndexOf(_)
+            | Lang::Add(_)
+            | Lang::Sub(_) => Type::Number,
+        }
     }
 
     fn eval<'a, F>(&'a self, cvec_len: usize, mut v: F) -> CVec<Self>
     where
         F: FnMut(&'a Id) -> &'a CVec<Self>,
     {
         match self {
-            Lang::Concat([a, b]) => map!(v, a, b => {
-                let a = a.to_slice().unwrap();
-                let b = b.to_slice().unwrap();
-                Some(Constant::Str(concat(a, b)))
-            }),
+            Lang::SVar(_) => vec![],
+            Lang::NVar(_) => vec![],
+            Lang::Lit(c) => vec![Some(c.clone()); cvec_len],
+            Lang::Concat([a, b]) => {
+                map!(v, a, b => Some(Constant::Str(SmallString::from_str(&format!("{}{}", a.to_str().unwrap(), b.to_str().unwrap())))))
+            }
             Lang::Replace([s, t, t2]) => map!(v, s, t, t2 => {
                 let s = s.to_str().unwrap();
                 let t = t.to_str().unwrap();
                 let t2 = t2.to_str().unwrap();
                 let out = s.replacen(t, t2, 1);
-                Some(Constant::Str(out.into_bytes().into()))
+                Some(Constant::Str(SmallString::from_str(&out)))
             }),
             Lang::At([a, i]) => map!(v, a, i => {
-                let a = a.to_slice().unwrap();
+                let a = a.to_smallstr().unwrap();
                 let i = i.to_num().unwrap();
-                if 0 <= i && i < (a.len() as _) {
-                    let i = i as usize;
-                    Some(Constant::Str(a[i..i+1].into()))
+                if i < 0 || i > (a.len() as _) {
+                    Some(Constant::Str(SmallString::default()))
                 } else {
-                    Some(Constant::Str(SmallStr::default()))
+                    let i = i as usize;
+                    Some(Constant::Str(SmallString::from(a.get(i..i+1).unwrap_or(""))))
                 }
             }),
             Lang::SubStr([a, i, len]) => map!(v, a, i, len => {
-                let a = a.to_slice().unwrap();
-                let n = a.len() as i64;
-                let i = i.to_num().unwrap();
-                let len = len.to_num().unwrap();
-                if 0 <= i && i < n && len > 0 {
-                    let j = (i + len).min(n) as usize;
-                    let i = i as usize;
-                    Some(Constant::Str(a[i..j].into()))
-                } else {
-                    Some(Constant::Str(SmallStr::default()))
+                let a = a.to_smallstr().unwrap();
+                let n = a.len() as usize;
+                let i = i.to_num().unwrap() as usize;
+                let len = len.to_num().unwrap() as usize;
+                let j = (i + len).min(n) as usize;
+                match a.get(i..j) {
```

Comment:
do we not need the condition  `0 <= i && i < n && len > 0`? Or does the `a.get(i..j)` do the right thing here?

Reply:
[get](https://doc.rust-lang.org/std/string/struct.String.html#method.get) should just return `None` if the indices aren't valid, so I don't think we need to check.